pkg/kubelet/volumemanager/reconciler/reconstruct_new.go

```diff
@@ -0,0 +1,241 @@
+				continue
+			}
+			klog.V(4).InfoS("Volume is marked as uncertain and added into the actual state", "pod", klog.KObj(volume.pod), "podName", volume.podName, "volumeName", volume.volumeName)
+		}
+		// If the volume has device to mount, we mark its device as mounted.
+		if gvl.deviceMounter != nil || gvl.blockVolumeMapper != nil {
+			deviceMountPath, err := getDeviceMountPath(gvl)
+			if err != nil {
+				klog.ErrorS(err, "Could not find device mount path for volume", "volumeName", gvl.volumeName)
+				continue
+			}
+			err = rc.actualStateOfWorld.MarkDeviceAsUncertain(gvl.volumeName, gvl.devicePath, deviceMountPath, "")
+			if err != nil {
+				klog.ErrorS(err, "Could not mark device is uncertain to actual state of world", "volumeName", gvl.volumeName, "deviceMountPath", deviceMountPath)
+				continue
+			}
+			klog.V(4).InfoS("Volume is marked device as uncertain and added into the actual state", "volumeName", gvl.volumeName, "deviceMountPath", deviceMountPath)
+		}
+	}
+}
+
+// cleanOrphanVolumes tries to clean up all volumes that failed reconstruction.
+func (rc *reconciler) cleanOrphanVolumes() {
+	if len(rc.volumesFailedReconstruction) == 0 {
+		return
+	}
+
+	for _, volume := range rc.volumesFailedReconstruction {
+		if rc.desiredStateOfWorld.VolumeExistsWithSpecName(volume.podName, volume.volumeSpecName) {
+			// Some pod needs the volume, don't clean it up and hope that
+			// reconcile() calls SetUp and reconstructs the volume in ASW.
+			klog.V(4).InfoS("Volume exists in desired state, skip cleaning up mounts", "podName", volume.podName, "volumeSpecName", volume.volumeSpecName)
+			continue
+		}
+		klog.InfoS("Cleaning up mounts for volume that could not be reconstructed", "podName", volume.podName, "volumeSpecName", volume.volumeSpecName)
+		rc.cleanupMounts(volume)
+	}
+
+	klog.V(2).InfoS("Orphan volume cleanup finished")
+	// Clean the cache, cleanup is one shot operation.
+	rc.volumesFailedReconstruction = make([]podVolume, 0)
+}
+
+// updateReconstructedDevicePaths tries to file devicePaths of reconstructed volumes from
+// node.Status.VolumesAttached. This can be done only after connection to the API
+// server is established, i.e. it can't be part of reconstructVolumes().
+func (rc *reconciler) updateReconstructedDevicePaths() {
+	klog.V(4).InfoS("Updating reconstructed devicePaths")
+
+	node, fetchErr := rc.kubeClient.CoreV1().Nodes().Get(context.TODO(), string(rc.nodeName), metav1.GetOptions{})
+	if fetchErr != nil {
+		// This may repeat few times per second until kubelet is able to read its own status for the first time.
+		klog.ErrorS(fetchErr, "Failed to get Node status to reconstruct device paths")
+		return
+	}
+
+	for _, volumeID := range rc.volumesNeedDevicePath {
+		for _, attachedVolume := range node.Status.VolumesAttached {
+			if volumeID != attachedVolume.Name {
+				continue
+			}
+			rc.actualStateOfWorld.UpdateReconstructedDevicePath(volumeID, attachedVolume.DevicePath)
```

Comment:
Is this racy with regular device path update that can happen with parallely running `mountOrAttachVolumes` function?

Reply:
That's why UpdateReconstructedDevicePath says "The ASW is updated only when the volume is still uncertain"